### ISSUE TITLE
Confusing behavior: non-target features are forecasted despite documentation suggesting otherwise

### ISSUE THREAD

**Issue by FerranAD** [ibm-granite/granite-tsfm#324]
### Description

I'm using the TimeSeriesPreprocessor and get_datasets() to prepare data for training a forecasting model with granite-tsfm. Based on the documentation, the intent is:

  - target_columns: variables to be forecasted

  - control_columns, observable_columns, conditional_columns: auxiliary features, with varying known/future availability

However, when inspecting the model outputs in the compute_metrics function:

  - p.predictions[0] and p.label_ids both have shape (batch_size, prediction_length, num_features)

  - num_features includes all feature types, not just the target

  - The non-target features do not match their corresponding values in label_ids, implying the model is forecasting those variables as well

This is confusing for a few reasons:

  - Documentation implies that only target_columns should be forecasted:
```python
class TimeSeriesPreprocessor(FeatureExtractionMixin):
    """
    A preprocessor for supporting time series modeling tasks
        Args:
            id_columns (List[str]): List of column names which identify different time series in a multi-time series input. Defaults to [].
            timestamp_column (Optional[str], optional): The name of the column containing the timestamp of the time series. Defaults to None.
            target_columns (List[str], optional): List of column names which identify the target channels in the input, these are the
                columns that will be forecasted. Defaults to [].
            observable_columns (List[str], optional): List of column names which identify the observable channels in the input.
                Observable channels are channels which we have knowledge about in the past and future. For example, weather
                conditions such as temperature or precipitation may be known or estimated in the future, but cannot be
                changed. Defaults to [].
    """
```

  An [example metric function used in other issues](https://github.com/ibm-granite/granite-tsfm/issues/273) computes RMSE over all features:

```python
def compute_metrics(p: EvalPrediction):
    return {"rmse": np.sqrt(np.square(p.predictions[0] - p.label_ids).mean())}
```

This would include error from non-target variables, potentially misleading users who expect loss/metrics to focus on `target_columns`.

### Expected behavior

If only target_columns are meant to be predicted:

-  Why does the model output predictions for other features?

-  Shouldn't p.predictions[0][:, :, non_target_idx] match p.label_ids[:, :, non_target_idx] exactly if, for some compatibility with transformers reason p.predictions[0] needs to contain values for all features.

### What would help clarify

  - Are all future features forecasted by design no matter if they are target or not?

  - If yes, isn't that contradictory with the documentation?

  - If no, is there a bug or unintended behavior in dataset construction?

**Comment by wgifford** [comment]
Hi @FerranAD thank you for your interest in TSFM. Can you provide a simple example of what you are doing? Of particular interest is how you are configuring / loading the model that produces the predictions you mentioned above. As an example, when we have exogenous features we need to specify the `prediction_channel_indices` and the `exogenous_channel_indices`:

```
model = TinyTimeMixerForPrediction.from_pretrained(
    "ibm-granite/granite-timeseries-ttm-r2",
    num_input_channels=tsp.num_input_channels,
    decoder_mode="mix_channel",  # exog:  set to mix_channel for mixing channels in history
    prediction_channel_indices=tsp.prediction_channel_indices,
    exogenous_channel_indices=tsp.exogenous_channel_indices, 
    ...
)
```
where `tsp` is a `TimeSeriesPreProcessor`. A complete example with exogenous features can be found here: https://github.com/ibm-granite-community/granite-timeseries-cookbook/blob/main/recipes/Time_Series/Bike_Sharing_Finetuning_with_Exogenous.ipynb

**Comment by FerranAD** [comment]
Thanks a lot for the quick response! The issue was indeed not specifying the indices.

Regarding the decoder mode, if I understand the documentation correctly, setting the decoder mode to "mix_channel" is a feature one uses when fine-tunning only, which seems to be confirmed by this warning I get when trying to use it with a pretrained model:

```
Some weights of TinyTimeMixerForPrediction were not initialized from the model checkpoint at ibm-granite/granite-timeseries-ttm-r2 and are newly initialized: ['decoder.decoder_block.mixers.0.channel_feature_mixer.gating_block.attn_layer.bias', 'decoder.decoder_block.mixers.0.channel_feature_mixer.gating_block.attn_layer.weight', 'decoder.decoder_block.mixers.0.channel_feature_mixer.mlp.fc1.bias', 'decoder.decoder_block.mixers.0.channel_feature_mixer.mlp.fc1.weight', 'decoder.decoder_block.mixers.0.channel_feature_mixer.mlp.fc2.bias', 'decoder.decoder_block.mixers.0.channel_feature_mixer.mlp.fc2.weight', 'decoder.decoder_block.mixers.0.channel_feature_mixer.norm.norm.bias', 'decoder.decoder_block.mixers.0.channel_feature_mixer.norm.norm.weight', 'decoder.decoder_block.mixers.1.channel_feature_mixer.gating_block.attn_layer.bias', 'decoder.decoder_block.mixers.1.channel_feature_mixer.gating_block.attn_layer.weight', 'decoder.decoder_block.mixers.1.channel_feature_mixer.mlp.fc1.bias', 'decoder.decoder_block.mixers.1.channel_feature_mixer.mlp.fc1.weight', 'decoder.decoder_block.mixers.1.channel_feature_mixer.mlp.fc2.bias', 'decoder.decoder_block.mixers.1.channel_feature_mixer.mlp.fc2.weight', 'decoder.decoder_block.mixers.1.channel_feature_mixer.norm.norm.bias', 'decoder.decoder_block.mixers.1.channel_feature_mixer.norm.norm.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
```

I was wandering then what is the role of exogenous features when performing zero-shot inference since each channel is worked on independently. Are they simply useless for zero-shot?

Thanks for your help and congrats on the project!

**Comment by FerranAD** [comment]
Thanks for pointing that out, is there any more information I can find regarding this arguments apart from that on the docstring?

```
enable_forecast_channel_mixing (`bool`, *optional*, defaults to `False`):
    Enable if we want to reconcile forecasts across all channels and also to enable exogenous infusion, if you have them.
fcm_gated_attn (`bool`, *optional*, defaults to `True`):
    Enable gated attention in forecast channel mixing block.
fcm_context_length (`int`, *optional*, defaults to `1):
    Surrounding context length to use. For Ex. If we want to consider 2 lag point before and after a data point, provide value 2 for `fcm_context_length`
fcm_use_mixer (`bool`, *optional*, defaults to `True`):
    Enable Mixing in forecast channel mixing block.
fcm_mix_layers (`int`, *optional*, defaults to 2):
    Number of mixer layers to use if fcm_use_mixer is enabled
fcm_prepend_past (`bool`, *optional*, defaults to `True`):
    Prepend last context for forecast reconciliation
fcm_prepend_past_offset  (`int`, *optional*, defaults to None):
```

**Comment by FerranAD** [comment]
Doing some research on the internet accompanied by the [auhtor's paper](https://arxiv.org/pdf/2401.03955) was enough to figure out how to play with this settings for my use case. Thanks for suggesting to consider this parameters!
